#install.packages('tidyverse')
library(tidyverse)
#install.packages('microseq')
library(microseq)
#install.packages('gtools')
library(gtools)


#This data is from a Thermo Ion S5 sequencer. 
#Metabarcoding data from a malaise trap ontario provincial parks dataset
data = readFastq('/home/cnuge/Documents/barcode_data/mBRAVE_raw_read_data/GMP-03299)CCDB-S5-0084)CBGMB-00030.fastq')

#on server
#data = readFastq('/home/cnugent/barcode_data/mBRAVE_raw_read_data/GMP-04500)CCDB-S5-0084)CBGMB-00030.fastq')

head(data)
names(data)

#Columns are header, sequence and quality.... can the quality column be leveraged?

#In FASTQ files, quality scores are encoded into a compact form, 
#which uses only 1 byte per quality value. In this encoding, 
#the quality score is represented as the character with an ASCII code equal to its value + 33.
#A Phred quality score is a measure of the quality of the identification of the nucleobases generated by automated DNA sequencing.
#Phred quality scores Q are defined as a property which is logarithmically related to the base-calling error probabilities P
#For example, if Phred assigns a quality score of 30 to a base, the chances that this base is called incorrectly are 1 in 1000.

#here is code for getting the quality information into a numeric vector, matches the positions in the 
#sequence strings, these could be passed in to a stastical model along with the 

phred_string = data$Quality[1]
phred_string

phred2numeric = function(phred_string){
	outvec = asc(unlist(strsplit(phred_string, "")))
	return(outvec - 33)
}

test = head(data)
test$qual_vec = lapply(test$Quality , function(x){phred2numeric(x)})
head(test)

#is using the phred scores as an input feature, would pass them in as another tensor to the input of the model

#if using artifical training data, would need to find the distribution of phred scores for inserted data, and the 
#distribution of phred scores of data next to deletions and then randomly assign phred scores based on these ditributions.
#^maybe best to go with real world data if using a phred layer on the input


#####
#try to dereplicate these reads

#get the OTU output from mBRAVE and look at the sequences of the OTUs, explore them for evidence
#of spurious OTUs resulting from sequencer errors.
